File: blob-module/ios/BlobModule.swift
import Foundation
import ExpoModulesCore

public class BlobModule: Module {
  public func definition() -> ModuleDefinition {
    Name("ExpoBlob")

    Class(Blob.self) {
      Constructor { (blobParts: [Either<String, Blob>]?, options: BlobOptions?) in
        let blobPartsProcessed: [BlobPart]? = blobParts?.map { part in
          if let part: String = part.get() {
            return part
          } else if let part: Blob = part.get() {
            return part
          }
          return ""
        }
        return Blob(blobParts: blobPartsProcessed ?? [], options: BlobOptions())
      }
      
      Property("size") { (blob: Blob) in
        blob.size
      }
      
      Property("type") { (blob: Blob) in
        blob.type
      }
      
      Property("endings") { (blob: Blob) in
        blob.options.endings
      }
      
      Function("slice") { (blob: Blob, start: Int?, end: Int?, contentType: String?) in
        blob.slice(start: start ?? 0, end: end ?? blob.size - 1, contentType: contentType ?? "")
      }
      
      Function("text") { (blob: Blob) in
        blob.text()
      }
    }
  }
}

File: blob-module/android/src/main/java/expo/modules/blobmodule/BlobModule.kt
package expo.modules.blobmodule

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class BlobModule : Module() {
    override fun definition() = ModuleDefinition {
        Name("ExpoBlob")

        Class(Blob::class) {
            Constructor() { blobParts: List<BlobPart>, options: BlobOptionsBag? ->
                var type = options?.type ?: BlobOptions().type
                var endings = options?.endings ?: BlobOptions().endings
                Blob(blobParts.internal(endings == EndingType.NATIVE), BlobOptions(type, endings))
            }

            Property("size") { blob: Blob ->
                blob.size
            }

            Property("type") { blob: Blob ->
                blob.options.type
            }

            Function("slice") { blob: Blob, start: Int?, end: Int?, contentType: String? ->
                var sliceStart: Int = start ?: 0
                var sliceEnd: Int = end ?: 0
                if (sliceStart < 0) {
                    sliceStart = blob.size + sliceStart
                }
                if (sliceEnd < 0) {
                    sliceEnd = blob.size + sliceEnd
                }
                blob.slice(sliceStart, sliceEnd, contentType ?: "")
            }

            AsyncFunction("bytes") { blob: Blob ->
                blob.text().toByteArray()
            }

            AsyncFunction("text") { blob: Blob ->
                blob.text()
            }
        }
    }
}
